calico.go

```diff
@@ -0,0 +1,265 @@
+	runtime.LockOSThread()
+
+	hostname, _ = os.Hostname()
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	if err := AddIgnoreUnknownArgs(); err != nil {
+		return err
+	}
+
+	// Unmarshall the network config, and perform validation
+	conf := NetConf{}
+	if err := json.Unmarshal(args.StdinData, &conf); err != nil {
+		return fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	// Allow the hostname to be overridden by the network config
+	if conf.Hostname != "" {
+		hostname = conf.Hostname
+	}
+
+	if err := ValidateNetworkName(conf.Name); err != nil {
+		return err
+	}
+
+	etcd, err := libcalico.GetKeysAPI(conf.EtcdAuthority, conf.EtcdEndpoints)
+	if err != nil {
+		return err
+	}
+
+	workloadID, orchestratorID, err := GetIdentifiers(args)
+	if err != nil {
+		return err
+	}
+
+	// Always check if there's an existing endpoint.
+	endpoint, err := libcalico.GetEndpoint(
+		etcd, libcalico.Workload{
+			Hostname:       hostname,
+			OrchestratorID: orchestratorID,
+			WorkloadID:     workloadID})
+
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "Calico CNI checking for existing endpoint. endpoint=%v\n", endpoint)
+
+	// Collect the result in this variable - this is ultimately what gets "returned" by this function by printing
+	// it to stdout.
+	var result *types.Result
+
+	// If running under Kubernetes then branch off into the kubernetes code, otherwise handle everything in this
+	// function.
+	if orchestratorID == "k8s" {
+		if result, err = k8s.CmdAddK8s(args, conf, hostname, etcd, endpoint); err != nil {
+			return err
+		}
+	} else {
+		// Not running in an environment that needs special casing.
+		profileID := conf.Name
+
+		if endpoint != nil {
+			// Don't create the veth or do any networking. Just update the profile on the endpoint. The profile will
+			// be created if needed during the profile processing step.
+
+			// There's an existing endpoint - no need to create another. Find the IP address from the endpoint
+			// and use that in the response.
+			fmt.Fprintf(os.Stderr, "Calico CNI appending networking profile=%s\n", profileID)
+			endpoint.ProfileIDs = append(endpoint.ProfileIDs, profileID)
+			result, err = CreateResultFromEndpoint(endpoint)
+			if err != nil {
+				return err
+			}
+		} else {
+			// There's no existing endpoint, so we need to do the following:
+			// 1) Call the configured IPAM plugin to get IP address(es)
+			// 2) Configure the calico endpoint
+			// 3) Create the veth, configuring it on both the host and container namespace.
+
+			// 1) run the IPAM plugin and make sure there's an IP address returned.
+			result, err = ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
+			if err != nil {
+				return err
+			}
+
+			// 2) Create the endpoint object
+			endpoint = &libcalico.Endpoint{ProfileIDs: []string{profileID}, OrchestratorID: orchestratorID,
+				WorkloadID: workloadID, Hostname: hostname, State: "active"}
+			if err = PopulateEndpointNets(endpoint, result); err != nil {
```

Comment:
Sounds like a very good idea, but none of the bundled CNI plugins have that behaviour. If we follow the principle of least surprise we should do the same as them. WDYT?